### PR TITLE
HPCC-18113 Unicode Implementation for EndsWith

### PR DESCRIPTION
**Pull request by dskaff**
- Adds test cases for EndsWith in ecllibrary/teststd/uni/TestEndsWith.ecl
- Adds documentation for EndsWith in ecllibrary/std/uni/Uni.ecl
- Adds code for EndsWith in unicodelib.cpp

Signed-off-by: David Skaff <david.skaff@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [x] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by dskaff** [comment]
@johnholt please code review.

File: plugins/unicodelib/unicodelib.cpp [hunk 1]
```diff
@@ -719,6 +720,56 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
...
+    UErrorCode errorCode = U_ZERO_ERROR;
+    U_STRING_DECL(C, "NFC", 3);
+    U_STRING_INIT(C, "NFC", 3);
+    U_STRING_DECL(D, "NFD", 3);
+    U_STRING_INIT(D, "NFD", 3);
+    U_STRING_DECL(KC, "NFKC", 4);
+    U_STRING_INIT(KC, "NFKC", 4);
+    U_STRING_DECL(KD, "NFKD", 4);
+    U_STRING_INIT(KD, "NFKD", 4);
+    if (form[2] == C[2])
```
Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: Much simpler to compare == 'C'.
See comments on the other PR

**Comment by dskaff** [comment]
@ghalliday please code review

File: plugins/unicodelib/unicodelib.cpp [hunk 2]
```diff
@@ -747,6 +748,19 @@ void normalizationFormCheck(UnicodeString & source, const char * form)
     }
 }
 
+bool endsWith(UnicodeString const & processed, UnicodeString const & suffix)
+{
+    if (processed.isEmpty() || suffix.isEmpty())
+    {
+        return false;
+    }
+    if (processed.endsWith(suffix) != 1)
```
Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: comparison with 1 is not valid.  Should be

    if (!processed.endsWith(suffix))

File: plugins/unicodelib/unicodelib.cpp [hunk 3]
```diff
@@ -747,6 +748,19 @@ void normalizationFormCheck(UnicodeString & source, const char * form)
...
+    {
+        return false;
+    }
+    return true;
+}
+
 bool startsWith(UnicodeString & processed, UnicodeString & prefix)
```
Comment (ghalliday) [inline plugins/unicodelib/unicodelib.cpp]: should have really been static bool as well.

**Pull request merged**